client/rpcserver/rpcserver.go

```diff
@@ -1,20 +1,405 @@
+	go func() {
+		<-ctx.Done()
+
+		if err := s.srv.Shutdown(context.Background()); err != nil {
+			// Error from closing listeners:
+			err = fmt.Errorf("HTTP server Shutdown: %v", err)
+			fmt.Fprintln(os.Stderr, err)
+		}
+	}()
+	log.Infof("RPC server listening on %s", s.listener.Addr())
+	if err := s.srv.Serve(s.listener); err != http.ErrServerClosed {
+		log.Warnf("unexpected (http.Server).Serve error: %v", err)
+	}
+	s.mtx.Lock()
+	for _, cl := range s.clients {
+		cl.Disconnect()
+	}
+	s.mtx.Unlock()
+
+	// Wait for market syncers to finish.
+	s.wg.Wait()
 	log.Infof("RPC server off")
 }
+
+// watchMarket watches the specified market for changes in orders. A fresh order
+// book and a quit function are returned on success. The quit function should be
+// called to unsubscribe the client from the market.
+func (s *RPCServer) watchMarket(cl *wsClient, dex string, base, quote uint32) (book *core.OrderBook, quit func(), err error) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	mktID := marketID(base, quote)
+	syncer, found := s.syncers[mktID]
+	if !found {
+		syncer, err = newMarketSyncer(s.ctx, &s.wg, s.core, dex, base, quote)
+		if err != nil {
+			return
+		}
+		s.syncers[mktID] = syncer
+	}
+	book = syncer.add(cl)
+	return book, func() {
+		syncer.remove(cl)
+	}, nil
+}
+
+// handleRequest sends the request to the correct handler function if able.
+func (s *RPCServer) handleRequest(req *msgjson.Message) *msgjson.ResponsePayload {
+	payload := new(msgjson.ResponsePayload)
+	if req.Route == "" {
+		log.Debugf("route not specified")
+		payload.Error = msgjson.NewError(msgjson.RPCUnknownRoute, "no route was supplied")
+		return payload
+	}
+
+	// Find the correct handler for this route.
+	h, ok := routes[req.Route]
+
+	if !ok {
+		log.Debugf("could not find route: %v", req.Route)
+		payload.Error = msgjson.NewError(msgjson.RPCUnknownRoute, "unknown route")
+		return payload
+	}
+
+	return h(s, req)
+}
+
+// parseHTTPRequest parses the msgjson message in the request body and creates a
+// response message.
+func (s *RPCServer) parseHTTPRequest(w http.ResponseWriter, req *msgjson.Message) {
+	payload := s.handleRequest(req)
+	resp, err := msgjson.NewResponse(req.ID, payload.Result, payload.Error)
+	if err != nil {
+		err = fmt.Errorf("error encoding response: %v", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		log.Error(err)
+		return
+	}
+	writeJSON(w, resp)
+}
+
+// authMiddleware checks incoming requests for authentication.
+func (s *RPCServer) authMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		auth := r.Header["Authorization"]
+		if len(auth) == 0 || s.authsha != sha256.Sum256([]byte(auth[0])) {
+			log.Warnf("authentication failure from ip: %s with auth: %s", r.RemoteAddr, auth)
+			w.Header().Add("WWW-Authenticate", `Basic realm="dex RPC"`)
+			http.Error(w, "401 Unauthorized.", http.StatusUnauthorized)
```

Comment:
I would generally advise to replace `"401 Unauthorized."` with `http.StatusText(http.StatusUnauthorized)` unless the message has details.